### ISSUE TITLE
git-crypt does not unlock files ...not able to view the file after unlock on MAC

### ISSUE THREAD

**Issue by beravelli** [AGWA/git-crypt#147]
I did lock the files using git-crypt lock and I see they are not readable...when I tried to do git-crypt unlock /path/to/key it exits with out any errors, but not able to see the content of the file..it is in encrypted format...can I decrypt the file with the key I have...

I need to disable the encryption for terraform to work.

FYI...I lost important state files and do not have back up. Please help

**Comment by cunneen** [comment]
I too have had issues on a mac. Something to do with the interaction of `gpg` and `gpg2` and their key details. 

[This comment](./99#issuecomment-270179200) on Issue #99 had the answer :

```bash
for i in "./.git-crypt/keys/default/0/*.gpg"
do
  filebase=`basename $i | sed -r "s/\.([^.]+)$//"`
  gpg2 -v --output "./.git-crypt/keys/default/0/${filebase}-decrypted.gpg" --decrypt "$i"
done
```

Then in your `.git-crypt/keys/default/0` folder, you should have a file with a `-decrypted.gpg` extension. Use that file as the argument to `git-crypt unlock <your_file_here>` .

**Comment by cunneen** [comment]
I also managed to get git-crypt working transparently again, by exporting the secret key from gpg2 and importing it into gpg:

```
gpg2 --export-secret-keys -a --output dummy-example-com-secret.gpg dummy@example.com
gpg2 --export -a --output dummy-example-com-public.gpg dummy@example.com
gpg --import dummy-example-com-secret.gpg 
# set the key to trusted
expect -c "spawn gpg --edit-key dummy@example.com trust quit; send \"5\ry\r\"; expect eof"
```

After that my `git-crypt unlock` works fine.

**Comment by jhnlsn** [comment]
I'm having the same issue, however, in my case 1 file out of about 20 files does not pass the smudge filter.  After that the rest of the files are not decrypted and the process fails.  All files were checked in at the same time, i'm not sure why one file got corrupted.

**Comment by m477r1x** [comment]
I've just found this because the same problem is happening to me. I've followed the instructions from @cunneen however i still cannot read the contents of the files. During the unlock process, it successfully reads the correct key in my gpg keychain and asks me for the passphrase, which is then accepts. It just doesn't seem to actually perform the unlock action. Did anybody ever discover anything more about this?

**Comment by mangatmodi** [comment]
Stuck on the same issue

**Comment by jethrogillgren** [comment]
Stuck on the same issue too, no error messages. Above not working for us.

**Comment by esantoro** [comment]
I'm stuck on the same issue: `git-crypt unlock` unlocks most of the files but not all of them.

I tried decrypting my key and use that to decrypt the repo but nothing. I've re-cloned the repo but nothing. I've exported my private/public key and reimported it but nothing.

i'm using git-crypt 0.6.0 and git version 2.17.1

**Comment by esantoro** [comment]
In my case, after digging into the sources, recompiling (with `-g` just in case), adding prints (`std::cout` really) here and there and verifying each stage of the process (unlock, git filters configuration, git filter invocation) etc etc i concluded that the code was sound enough and the problem was not really in the code. 

By adding some `std::cout << stuff << std::endl;` here and there (specifically in the smudge function -- command?) it seemed more evident that some paths from the `.gitattributes` were being picked for `smudge`-ing and some not.

So I made the paths regex for the missing files less "smart" (aka less regexps, less `**`, more explicit and repeated paths) and it started picking up the files.

I wrote all this not to brag about this, but for other people (this looks like a common problem) to double-check the path regexes in the `.gitattribute` file. 

I just checked on one of the VMs where i used for testing and the `.gitattributes` fix works there too.

**TL;DR:** double-check the paths in your `.gitattributes`. try making it more explicit (less regexes)

**Comment by TheTravel3r** [comment]
## Reproducing on macOS with git-crypt 0.8.0 (March 2026)

**Environment:** macOS 15.4, git-crypt 0.8.0 (Homebrew), symmetric key (no GPG)

**Reproduction:**
1. Repo with `* filter=git-crypt diff=git-crypt` in `.gitattributes`
2. Clone repo → files are encrypted blobs on disk (expected before unlock)
3. Run `git-crypt unlock /path/to/symmetric-key` → exits 0, no error
4. Files remain encrypted on disk. `git-crypt status` still shows `encrypted:`
5. `git checkout -- .` does NOT re-trigger the smudge filter

**Root cause:** `git-crypt unlock` stores the key in `.git/git-crypt/keys/default` but does not force git to re-checkout files through the smudge filter. Git considers the working tree files "unchanged" (same content as index), so `checkout` is a no-op.

**Confirmed:** Running `git-crypt smudge` manually on each file works perfectly:
```bash
"git-crypt" smudge < encrypted_file > decrypted_file && mv decrypted_file encrypted_file
```

**Workaround:**
```bash
# After git-crypt unlock, force re-decrypt all encrypted files:
git-crypt status 2>/dev/null | grep "encrypted:" | sed 's/.*encrypted: //' | while read f; do
  if [ -f "$f" ]; then
    "git-crypt" smudge < "$f" > "$f.dec" && mv "$f.dec" "$f"
  fi
done
```

**Proposed fix:** After storing the key, `git-crypt unlock` should either:
1. Run `git checkout --force -- .` to force smudge filter re-application, or
2. Iterate encrypted files and pipe each through `smudge` (as the workaround does)

This has been open since 2018 and affects many users. Happy to submit a PR if there's interest.
